### PR TITLE
CRA in flowbased

### PR DESCRIPTION
**Pull request by terriervik**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Feature

**What is the new behavior (if this is a feature change)?**
Curative remedial actions are now handled in flowbased computation, the same way as PRAs.

### CODE REVIEW

**Comment by terriervik** [comment]
This branch is based on https://github.com/farao-community/farao-core/pull/401.

File: data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/CracResultUtil.java [hunk 1]
```diff
@@ -41,11 +41,11 @@ public static void applyRemedialActionsForState(Network network, Crac crac, Stat
         CracResultExtension cracExtension = crac.getExtension(CracResultExtension.class);
         ResultVariantManager resultVariantManager = crac.getExtension(ResultVariantManager.class);
         if (resultVariantManager != null && cracExtension != null) { // Results from RAO
-            LOGGER.debug("Remedial Actions selected from RAO results.");
+            LOGGER.debug("Preventive Remedial Actions selected from RAO results.");
```
Comment (murgeyseb) [inline data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/CracResultUtil.java:44]: This message should depend from state ?

File: data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/CracResultUtil.java [hunk 2]
```diff
@@ -41,11 +41,11 @@ public static void applyRemedialActionsForState(Network network, Crac crac, Stat
...
             String cracVariantId = findPostOptimVariant(resultVariantManager, cracExtension);
             applyRemedialActionsForState(network, crac, cracVariantId, state);
         } else { // Apply all RAs from CRAC
-            LOGGER.debug("No RAO results found. All Remedial Actions from CRAC are applied.");
+            LOGGER.debug("No RAO results found. All Preventive Remedial Actions from CRAC are applied.");
```
Comment (murgeyseb) [inline data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/CracResultUtil.java:48]: This message should depend from state ?

File: flowbased-computation/flowbased-computation-impl/src/main/java/com/farao_community/farao/flowbased_computation/impl/FlowbasedComputationImpl.java [hunk 3]
```diff
@@ -67,6 +69,33 @@ public String getVersion() {
         SystematicSensitivityResult result = systematicSensitivityInterface.run(network);
         FlowbasedComputationResult flowBasedComputationResult = new FlowbasedComputationResultImpl(FlowbasedComputationResult.Status.SUCCESS, buildFlowbasedDomain(crac, glsk, result));
 
+        AtomicReference<String> curativeState = new AtomicReference<>("");
```
Comment (murgeyseb) [inline flowbased-computation/flowbased-computation-impl/src/main/java/com/farao_community/farao/flowbased_computation/impl/FlowbasedComputationImpl.java:72]: Avoid usage of Atomic reference by replacing Stream.forEach() method by a standard for-each loop:
```java
for (Instant instant : crac.getInstants()) {
    if (instant.getSeconds() > seconds) {
        seconds = (double) instant.getSeconds();
        curativeState = instant.getId();
    }
}
```

Do the same kind of modification for the loop below

**Review by murgeyseb — CHANGES_REQUESTED** [review]

File: data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/CracResultUtil.java (hunk 1 above)
Comment (terriervik) [reply data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/CracResultUtil.java]: Done

File: data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/CracResultUtil.java (hunk 2 above)
Comment (terriervik) [reply data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/CracResultUtil.java]: Done

File: flowbased-computation/flowbased-computation-impl/src/main/java/com/farao_community/farao/flowbased_computation/impl/FlowbasedComputationImpl.java (hunk 3 above)
Comment (terriervik) [reply flowbased-computation/flowbased-computation-impl/src/main/java/com/farao_community/farao/flowbased_computation/impl/FlowbasedComputationImpl.java]: Done, and I moved this to an independent method too.

**Comment by terriervik** [comment]
The question is to determine how to handle both OnOutage and AfterCra in the results object.

**Comment by terriervik** [comment]
> The question is to determine how to handle both OnOutage and AfterCra in the results object.

I've updated the PR, a new attribute (instantId) is added to the dataMonitoredBranches, to handle this problem.

**Comment by terriervik** [comment]
The Cucumber tests are successful, for a simple 3 nodes network.

**Comment by terriervik** [comment]
@baptsegu I've pushed a version with your correction.

File: flowbased-computation/flowbased-computation-impl/src/main/java/com/farao_community/farao/flowbased_computation/impl/FlowbasedComputationImpl.java [hunk 4]
```diff
@@ -36,6 +37,9 @@
 public class FlowbasedComputationImpl implements FlowbasedComputationProvider {
 
     static final String INITIAL_STATE_WITH_PRA = "InitialStateWithPra";
+    String onOutageInstantId = "";
+    String afterCraInstantId = "";
```
Comment (baptsegu) [inline flowbased-computation/flowbased-computation-impl/src/main/java/com/farao_community/farao/flowbased_computation/impl/FlowbasedComputationImpl.java:41]: both these parameters can be private

**Review by baptsegu — APPROVED** [review]
Ok for me, thanks for the modification of the curative sensitivity analysis. Just one very small comment.

File: flowbased-computation/flowbased-computation-impl/src/main/java/com/farao_community/farao/flowbased_computation/impl/FlowbasedComputationImpl.java (hunk 4 above)
Comment (baptsegu) [reply flowbased-computation/flowbased-computation-impl/src/main/java/com/farao_community/farao/flowbased_computation/impl/FlowbasedComputationImpl.java:41]: same for the four attributes of this class actually

**Review by baptsegu — DISMISSED** [review]
thanks for the changes

**Comment by baptsegu** [comment]
@murgeyseb if it's ok for you we can merge it.

File: data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/CracResultUtil.java [hunk 5]
```diff
@@ -41,11 +41,13 @@ public static void applyRemedialActionsForState(Network network, Crac crac, Stat
         CracResultExtension cracExtension = crac.getExtension(CracResultExtension.class);
         ResultVariantManager resultVariantManager = crac.getExtension(ResultVariantManager.class);
         if (resultVariantManager != null && cracExtension != null) { // Results from RAO
-            LOGGER.debug("Remedial Actions selected from RAO results.");
+            String raFromRaoMessage = String.format("Remedial Actions selected from RAO results for state %s.", state.getId());
+            LOGGER.debug(raFromRaoMessage);
```
Comment (murgeyseb) [inline data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/CracResultUtil.java:45]: Replace by:
```
LOGGER.debug("Remedial Actions selected from RAO results for state {}.", state.getId());
````

File: data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/CracResultUtil.java [hunk 6]
```diff
@@ -41,11 +41,13 @@ public static void applyRemedialActionsForState(Network network, Crac crac, Stat
...
             String cracVariantId = findPostOptimVariant(resultVariantManager, cracExtension);
             applyRemedialActionsForState(network, crac, cracVariantId, state);
         } else { // Apply all RAs from CRAC
-            LOGGER.debug("No RAO results found. All Remedial Actions from CRAC are applied.");
+            String allRaMessage = String.format("No RAO results found. All Remedial Actions from CRAC are applied for state %s.", state.getId());
+            LOGGER.debug(allRaMessage);
```
Comment (murgeyseb) [inline data/crac/crac-result-extensions/src/main/java/com/farao_community/farao/data/crac_result_extensions/CracResultUtil.java:50]: Replace by:
```
LOGGER.debug("No RAO results found. All Remedial Actions from CRAC are applied for state {}.", state.getId());
```

File: flowbased-computation/flowbased-computation-impl/src/main/java/com/farao_community/farao/flowbased_computation/impl/FlowbasedComputationImpl.java [hunk 7]
```diff
@@ -35,7 +36,10 @@
 @AutoService(FlowbasedComputationProvider.class)
 public class FlowbasedComputationImpl implements FlowbasedComputationProvider {
 
-    static final String INITIAL_STATE_WITH_PRA = "InitialStateWithPra";
+    private static final String INITIAL_STATE_WITH_PRA = "InitialStateWithPra";
+    private String onOutageInstantId = "";
```
Comment (murgeyseb) [inline flowbased-computation/flowbased-computation-impl/src/main/java/com/farao_community/farao/flowbased_computation/impl/FlowbasedComputationImpl.java:40]: set null as default value (and null check when needed)

File: flowbased-computation/flowbased-computation-impl/src/main/java/com/farao_community/farao/flowbased_computation/impl/FlowbasedComputationImpl.java [hunk 8]
```diff
@@ -35,7 +36,10 @@
...
+    private String afterCraInstantId = "";
+    private final Logger logger = LoggerFactory.getLogger(FlowbasedComputationImpl.class);
```
Comment (murgeyseb) [inline flowbased-computation/flowbased-computation-impl/src/main/java/com/farao_community/farao/flowbased_computation/impl/FlowbasedComputationImpl.java:42]: static logger (and call it LOGGER)

**Review by murgeyseb — CHANGES_REQUESTED** [review]

**Comment by terriervik** [comment]
@murgeyseb I corrected what you told me.

**Review by murgeyseb — APPROVED** [review]

**Pull request merged**